added layout and logic

Filosuf opened the pull request:
домашняя работа к спринту 4
не добавлен экран ошибки, не понятно в каких случаях его вызывать
По макетам тоже не всё идеально, предоставлен макет только для одного экрана, а приложение просят сделать на несколько.

Filosuf committed:
added layout and logic

ABashkirova commented on MovieQuiz/Helpers/UIColor+Extensions.swift:
```diff
@@ -1,4 +1,30 @@
+        )
+    }
+
+    enum YPTheme {
+        static var black: UIColor { return UIColor(rgb: 0x1A1B22) }
+        static var gray: UIColor { return UIColor(rgb: 0xE6E8EB) }
+        static var green: UIColor { return UIColor(rgb: 0x60C28E) }
+        static var red: UIColor { return UIColor(rgb: 0xF56B6C) }
+        static var white: UIColor { return UIColor(rgb: 0xFFFFFF) }
+        static var background: UIColor { return UIColor(red: 26, green: 27, blue: 34, alpha: 0.6) }
+    }
```
Добавлять функциональность для работы с hex в этой задаче избыточно. 
В задачнике было показано как использовать цвета, которые хранятся в assets-каталоге. Давайте обновим код на более нативный вариант, ведь он не требует дополнительного кода. 

У такого подхода есть классные преимущества, например, цвет в ассетах легко поддерживает светлую и темную темы из коробки.

ABashkirova commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,251 @@
-        super.viewDidLoad()
+struct QuizeQuestion {
+    // MARK: - Properties
+    let image: String
```
Я думаю, все таки это больше imageName, чем image. Как вы считаете?

ABashkirova commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,251 @@
+    private var gameCount = 0
+    private var averageAccuracy = 0.0
+    private var recordDate = Date()
+    private lazy var backgroundView: UIView = {
+        let backgroundView = UIView(frame: self.view.frame)
+        backgroundView.backgroundColor = .YPTheme.background
+        backgroundView.alpha = 0
+        return backgroundView
+    }()
```
Это очень странная вьюшка. По названию – выглядит как backgroundView, а используется как временный overlay. Может стоит тогда ее переименовать?

ABashkirova commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,251 @@
+    private func showNextQuestionOrResults() {
+        if currentQuestionIndex == numberOfQuestionsInGame - 1 {
+            // показать результат квиза
+            gameCount += 1
+            allCorrectAnswer += currentCorrectAnswer
+            var alertTitle = "Этот раунд окончен!"
+            if currentCorrectAnswer >= recordCorrectAnswer {
+                alertTitle = "Новый рекорд!"
+                recordCorrectAnswer = currentCorrectAnswer
+                recordDate = Date()
+            }
+
+            if currentCorrectAnswer == numberOfQuestionsInGame {
+                alertTitle = "Поздравляем. Лучший результат!"
+            }
+            let recordDateString = dateToString(date: recordDate)
+            averageAccuracy = Double(allCorrectAnswer * 100) / Double(numberOfQuestionsInGame * gameCount)
+            let resultTemp = QuizeResultsViewModel(
+                title: alertTitle,
+                text: """
+                Ваш результат:\(currentCorrectAnswer)/\(numberOfQuestionsInGame)
+                Количество сыграных квизов: \(gameCount)
+                Рекорд: \(recordCorrectAnswer)/\(numberOfQuestionsInGame) \(recordDateString)
+                Средняя точность: \(String(format: "%.02f", averageAccuracy))%
+                """,
+                buttonText: "Начать новую игру"
+            )
```
Этот кусок довольно тяжело воспринимать. Попробуйте поделить на шаги: 
1) обновить стейт,
2) сконфигурировать модель результата 

Посмотрите на блок с else: он ведь довольно компактный получился. Его гораздо легче воспринимать.

ABashkirova commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,251 @@
+    private func startGame() {
+        currentCorrectAnswer = 0
+        currentQuestionIndex = 0
+        questions.shuffle()
```
like!

ABashkirova commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,251 @@
+            preferredStyle: .alert) // preferredStyle может быть .alert или .actionSheet
+
+        // создаём для него кнопки с действиями
+        let action = UIAlertAction(title: result.buttonText, style: .default) { _ in
+            // убираем затемнение фона
+            UIView.animate(withDuration: 0.25) {
+                self.backgroundView.alpha = 0
+            }
+            self.backgroundView.removeFromSuperview()
+            self.startGame()
+        }
```
Здесь используется замыкание. В котором очень легко сделать ошибку, которую очень тяжело искать, когда приложение разрастается. Ваше замыкание захватывает сильную ссылку на класс. Это может приводить к утечке памяти, вы изучите это в ближайших спринтах. Однако в уроках упоминалось, что здесь лучше использовать weak self. Хотя в наших условиях утечки и не произойдет, давайте воспринимать это как хорошую привычку :)

ABashkirova commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,251 @@
+            self.backgroundView.alpha = UIColor.YPTheme.background.cgColor.alpha
+        }
+        // показываем всплывающее окно
+        self.present(alert, animated: true, completion: nil)
```
```suggestion
        present(alert, animated: true, completion: nil)
```

Используйте короткую версию, каждый раз, когда это возможно

ABashkirova commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,251 @@
+    }
 
+    private func convert(model: QuizeQuestion) -> QuizeStepViewModel {
+        let image = UIImage(named: model.image) ?? UIImage(systemName: "exclamationmark.icloud.fill")!
```
`UIImage(systemName: "exclamationmark.icloud.fill")!`

я не против использования force unwrap для данного случая. Однако предлагаю вынести этот код в константу (notAvailableImage), так будет более читаемо :)

ABashkirova commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,251 @@
- Настоящий рейтинг: 5,1
- Вопрос: Рейтинг этого фильма больше чем 6?
- Ответ: НЕТ
+    private func buttonTapped(response: Bool) {
```
buttonTapped – это больше про произошедшее действие, которое вы уже отловили в IBAction. Далее по коду здесь скорее обрабатывается результат нажатой кнопки. Давайте использовать что-то вроде handleAnswer

ABashkirova commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,251 @@
- Вопрос: Рейтинг этого фильма больше чем 6?
- Ответ: НЕТ
- */
+    private func dateToString(date: Date) -> String {
+        let formatter = DateFormatter()
+        formatter.dateFormat = "dd.MM.YY HH:mm"
+        return formatter.string(from: date)
+    }
```
Вам предлагается уже готовое решение с форматeром в Date+extensions. Давайте использовать его. 

Обратите внимание, форматер в экстеншене используется глобально(=статически). Так сделано не случайно, ведь создание форматтера [дорогая операция](https://sarunw.com/posts/how-expensive-is-dateformatter/). Ничего страшного не происходит, однако это занимает время каждый раз, когда вы показываете результат. Поэтому стоит сэкономить здесь :)

ABashkirova commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,251 @@
+        // добавляем в алерт кнопки
+        alert.addAction(action)
+
+        // затемнение фона
+        view.addSubview(backgroundView)
+        UIView.animate(withDuration: 0.25) {
+            self.backgroundView.alpha = UIColor.YPTheme.background.cgColor.alpha
+        }
```
Здорово, что не забыли про этот момент! Его достаточно легко не заметить на макетах :)

ABashkirova commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,251 @@
+        } else {
+            showCorrectAnswer(response: false)
+        }
+        for button in buttons {
+            button.isEnabled = false
+        }
```
Отлично, что обработали ситуацию! Этот код можно это вынести в функцию и переиспользовать

ABashkirova reviewed (changes requested):
Привет, Евгений! Отличная работа, но нужно немного доработать. 

Вижу твой комментарий в PR. 

> не добавлен экран ошибки, не понятно в каких случаях его вызывать
По макетам тоже не всё идеально, предоставлен макет только для одного экрана, а приложение просят сделать на несколько. 

Действительно, экран ошибки пока не нужен. Скоро вы будете расширять возможности квиза и он обязательно еще понадобится. 

Про предоставленные макеты. Действительно, очень удобно, когда дизайнер готовит макеты тщательно показывая как "работает" экран. Однако зачастую это не так. Грамотный софтскильный разработчик проверит макеты на недостатки и понятность, и при необходимости запросит больше деталей у дизайнера. А бывает и иначе - дизайнер отработал "по заказу", а компания не заложила в бюджет правки, и тогда iOS разработчику приходится использовать знания [HIG](https://developer.apple.com/design/human-interface-guidelines/guidelines/overview/), свою насмотренность и личный опыт, чтобы решить проблему самостоятельно. 

Однако большинство неоднозначностей в макетах задания разрешено в уроках к спринту, там дан необходимый контекст, чтобы сверстать экран корректно.  

При этом небольшие отличия в трактовке дизайна будут приняты, при условии, что в остальном приложение отображается корректно :)

Filosuf committed:
Исправления в коде по замечаниям полученным при ревью в PR

Filosuf committed:
Storibord переделан на StackView.
Исправлены замечания полученные на платформе Яндекс.Практикум
- картинка сжимается
- удалены  лишние свойства кнопок
- количество вопросов исправлено на 10

Filosuf committed:
refactoring

Filosuf committed:
added Autoshrink 0.5 for Questions Label

Filosuf committed:
fixed warnings "swift linter"
